### ISSUE TITLE
Configuration of serializers

### ISSUE THREAD

**Issue by rubengrill** [awesto/django-shop#486]
Here some thoughts about making (all) the shop serializers configurable:

1. Currently only `CustomerSerializer` and `OrderItemSerializer` are configurable, should we add a configuration setting for all serializers? In a project we needed to add a field `minimum_order_value_reached` to the `CartSerializer`, but we weren't able to do that easily on server side, so we had to parse the price string on client side to do that.
2. Both are by default not compatible with the base `Customer` and `OrderItem` models
  - `CustomerSerializer` serializes `salutation`, which is not contained in base Customer
  - For this it uses `get_salutation_display` which is not contained at all in the source code?
  - `OrderItemSerializer` serializes `quantity` which is not contained in base OrderItem
    This field is sth I really miss in the base `OrderItem`, should we add this field there by default?
3. We have two different approaches on how to configure serializers:
  - Via configuration setting (`SHOP_CUSTOMER_SERIALIZER`, ...)
  - Via subclassing a base serializer (`ProductSummarySerializerBase`)
  Should we change that to just use the configurations (and keep the subclass strategy for backward compatibility)?
4. Structure of serializers gets complicated with `defaults`, `bases` and `serializers`. 
  I think that simply exporting the configured serializer in `serializers` is problematic:
  If a shop extends the `CustomerSerializer` from `serializers` and then changes the setting of the customer serializer, the extending serializer now extends a different class.
  I would suggest a different strategy:
  - There is `serializers` defining default serializers that are already working serializers for the base models. They don't change when the configured serializers change.
  - There is `defaults` defining default serializers that can be used when the default models are used (not automatically, we should state in the documentation, that if you use the default models, then you should also configure the default serializers)
  - The views are overriding `get_serializer_class` to return the configured serializers

**Comment by jrief** [comment]
Just an explanation from a historical perspective:

The first and only configurable serializer was the ``ProductSummarySerializer``. It is required to display the smallest common denominator of content from various (even polymorphic) products. This must be implemented by the merchant.

I then realized, that it makes much more sense to use the ``fields`` attribute in the ``Meta`` class of serializers, rather than the ``exclude`` attribute. This however required the merchant implementation to somehow configure the serializers for ``CustomerSerializer``, ``OrderItemSerializer``, etc.

In Django we have three options to configure them:

1.  Do it in ``settings.py``. This is how it currently is (temporarily, quick & dirty) implemented, but I prefer to keep these kind of settings inside the code, rather than the configuration. I really disliked the model configurations in django-SHOP 0.2.
2. Create a self-registering meta-class. This is how the ``ProductSummarySerializer`` currently is implemented. It also is analogous to how shop models are registered now, using the "deferred pattern". This pattern It also used by Haystack and is considered good practice in Python *).
3. Use an explicit registry, similar to how it works in Django-Admin, or the django-CMS plugin system.

My personal opinion is to use the 2nd proposal. I oppose the 1st proposal, but can be convinced to chose the 3rd one.

So lets debate about 2nd vs. 3rd. I prefer the 2nd because a ``CustomerSerializer``, ``OrderItemSerializer`` and a ``ProductSummarySerializer`` must be implemented anyway. This is not like the ModelAdmin-, or the CMS-plugin classes, which optionally can be registered. Those serializers *must be* implemented by the merchant once and only once. If they are not implemented or implemented twice, the framework raises an exception. It also is possible to inherit from such a serializer, then the final serializer class is used, replacing the intermediate one. But it is not possible to inherit twice from, say ``BaseCustomerSerializer``.

Hence in my opinion, we should proceed with proposal 2.

@rfleschenberg disagreed on this, he prefers proposal 3.

@rubengrill 
> Structure of serializers gets complicated with defaults, bases and serializers.

I fully agree on that. I prefer to refactor the serializers into a new module named ``shop.serializers`` containing ``bases`` and a submodule ``defaults``. We then have a folder structure, analogous to that in ``shop.models`` and ``shop.admin``.

> There is defaults defining default serializers that can be used when the default models are used

Yes, similar to how models are implemented now. If the merchant implementation imports that serializer class into any of it's Python modules, then it is auto-registered by the framework.

However, I wanted to wait for this refactoring step until version 0.9.5, because I need an intermediate version 0.9.4 for my current merchant implementations.

*) http://www.effectivepython.com/2015/02/02/register-class-existence-with-metaclasses/

**Comment by rubengrill** [comment]
I agree that for the shop models it feels strange to configure them in the settings.
Mainly because setting up models affects a lot of things like migrations, relations, ..., so it should not be possible to simply switch models by configuration. Another reason might be, that the models already express their relations, so it is somehow redundant to specify all the models again in the configuration.

For serializers (or forms like in allauth) I don't find configuration in settings bad, also because it is such a common pattern in django (context processors, middlewares, auth backends, staticfile finders, storage, ...).
In fact, for me it is the first thing I would check when I want to change sth within a third party app, I would first check the available settings before digging into the documentation or even source code.

Another advantage of this approach is that settings can be changed per deployment via environment variables, sth that could not be done so easily with implicit or explicit registration of serializers (though I didn't have this requirement so far).

When choosing between implicit or explicit registration I would also prefer the explicit one.
This is, because there is no such concept as abstract models for serializers, so an implicit registration would not allow developers to have their own base serializer with common stuff that they might use in every of their projects.
Maybe it is possible to check if a second registration is a subclass of a previously registered serializer, but this could get complex quickly. The base class could be changed to be a mixin, but it should be the choice of the developers, if they want to use a mixin or have a base class instead.
Another scenario is that developers want to deploy a new serializer to their staging environment for testing, so there would be two different serializers extending the same base serializer.
Also, an implicit registration is somehow magic, I remember that when I first had to implement the `ProductSummarySerializer`, I was totally wondering how django-SHOP knows that it should use this serializer now.

For the explicit registration, I'm not sure how the registration should look like.
Can we rely on the fact that the serializer has to extend a base serializer from the shop, so that we are able to realize that a given serializer should replace an existing serializer just because it is a subclass of it?
Or should it be more like when registering a `ModelAdmin`, where both the `ModelAdmin` and the `Model` have to be given.
This way we would allow developers to completely replace a serializer implementation.

**Comment by jrief** [comment]
So, you are arguing in favor of my 1st proposal. I see two advantages: 1) we can configure default serializers, so that the merchant only has to adopt the configuration, if he wants to use something implemented by himself. 2) it would be quite easy to switch site, with different client-side REST implementations. For this however, I believe it should also be possible to use different URL patterns and Views.

A major problem in django-SHOP version 0.2 with models configured though ``settings.py``, was the handling of circular dependencies (Foreign keys pointing onto each other). If one day every serializer is configurable through settings, do we then run into the same kind of problems?

@rfleschenberg what's your opinion about this?

**Comment by rfleschenberg** [comment]
My votes are:
* settings: +0
* metaclass: -1
* registry: +1

Implicit registration by metaclass is too much magic for my taste, and it imposes unnecessary restrictions on the user, as Ruben has pointed out. Metaclass code is also harder to understand for contributors, and harder to test. IMHO metaclasses should only be used when necessary.

Ruben makes a good point about being able to change the configured serializers via environment variables in settings. On the other hand, with an explicit registry, it wouldn't be too hard to write user code that examines a custom setting and conditionally registers / unregisters a serializer based on that .

A potential advantage of an explicit registry would be that you can change serializers at runtime. I don't see a good use case for that at the moment, though.



If we go with 3 (explicit registry), then I think we should do it similarly to `django.contrib.admin`. My first idea would be something like this:

```python
from shop import serializers

serializers.unregister('PRODUCT_SUMMARY')

# If there already is a serializer registered, this should throw an exception.
serializers.register('PRODUCT_SUMMARY', MyCustomProductSummarySerializer)

# Maybe also provide this for convenience?
serializers.register('PRODUCT_SUMMARY', MyCustomProductSummarySerializer, 
                      overwrite_existing=True)
```

Maybe we could also combine 1 and 3? Expose an explicit registry API, and in `shop.apps.AppConfig.setup()`, register some serializers based on settings and defaults?

**Comment by rubengrill** [comment]
I don't feel very comfortable with the explicit registration after some thoughts now.
It requires that a developer does/knows three things about the API:
What is the module/name of the register/unregister function, what is the key/constant of the serializer to override, and also which serializer they should override (assuming that it is an edge case to provide a completely custom serializer).

And also, I think that a registry fits better, when multiple classes are allowed to be registered (like CMS plugins) or if the registry should bind two independent classes (like model admin, remember that a ModelAdmin could be reused, so it is not strictly bound to a Model. Apart from a serializer, which will map to exactly one base serializer).

Thats why I would prefer the implicit over the explicit registration, even if I also see some drawbacks with the implicit as well, but here simplicity is just most important to me.

- settings: +1
- metaclass: 0
- registry: -1

It seems we have a draw :)

**Comment by jrief** [comment]
Let's summarize ...

@jrief (1st to vote)
* settings: -1
* metaclass: +1
* registry: 0

@rfleschenberg (2nd to vote) 
* settings: +0
* metaclass: -1
* registry: +1

@rubengrill (3rd to vote)
* settings: +1
* metaclass: 0
* registry: -1

so all options got 0 points, no winner, no looser :)

Since @rubengrill added some valuable thoughts about this, I have to admit that my previous decline on the *settings* approach no longer is valid. Therefore, unless @rfleschenberg changes his mind, that's the winner. So, let's proceed with it.

**Comment by rfleschenberg** [comment]
Yup, let's go with settings.

**Comment by jrief** [comment]
Just one additional note. Since our models have circular dependencies on each other, and since our serializers reflect these models somehow, we may have circular dependencies on the serializers too.

This was one of the major headaches I had with django-SHOP version 0.2. Therefore, if we ever run into a similar problem with circular serializer dependencies, I'd like to retain the right to return to the *metaclass* pattern.

**Comment by rubengrill** [comment]
ok super!

@jrief Sure, that sounds reasonable 

Should I start implementing it then in a PR? I would make a branch on top of the `app_settings` branch, because I think that this could be important to avoid circular dependencies.
Maybe for version 0.9.5 together with the settings then.

**Comment by jrief** [comment]
Serializers have been refactored to use configuration settings.
This is part of the branch ``releases/0.10.x`` and waiting for review in #491
